pkg/authz/check/service.go

```diff
@@ -405,92 +314,395 @@ func (svc CheckService) CheckMany(ctx context.Context, authInfo *service.AuthInf
+}
+
+func NewPipeline(maxServiceConcurrency int, maxSubtaskConcurrency int) *pipeline {
+	return &pipeline{
+		serviceSemaphore: make(chan struct{}, maxServiceConcurrency),
+		subtaskSemaphore: make(chan struct{}, maxSubtaskConcurrency),
+	}
+}
+
+func (p *pipeline) AcquireServiceLock() {
+	p.serviceSemaphore <- struct{}{}
+}
+
+func (p *pipeline) ReleaseServiceLock() {
+	<-p.serviceSemaphore
+}
+
+func (p *pipeline) AnyOf(ctx context.Context, parentResultC chan<- result, tasks []func(execCtx context.Context, resultC chan<- result)) {
+	p.execTasks(ctx, parentResultC, tasks, func(res result, isLastExpected bool) (*result, bool) {
+		// Short-circuit - pick this result if it's a match
+		if res.Matched {
+			return &res, true
+		}
+		// Last result AND it's not a match due to prev condition -> return not matched
+		if isLastExpected {
+			return &result{
+				Matched:      false,
+				DecisionPath: res.DecisionPath,
+				Err:          nil,
+			}, true
 		}
+		// Not a match, keep looking
+		return nil, false
+	})
+}
 
-		if match {
-			return nil
+func (p *pipeline) AllOf(ctx context.Context, parentResultC chan<- result, tasks []func(execCtx context.Context, resultC chan<- result)) {
+	p.execTasks(ctx, parentResultC, tasks, func(res result, isLastExpected bool) (*result, bool) {
+		// Short-circuit - return not matched if any sub-result is not matched
+		if !res.Matched {
+			return &res, true
+		}
+		// Last result AND it's a match due to prev condition -> return matched
+		if isLastExpected {
+			return &result{
+				Matched:      true,
+				DecisionPath: res.DecisionPath,
+				Err:          nil,
+			}, true
 		}
+		// Keep looking
+		return nil, false
+	})
+}
 
-		match = false
-		return nil
+func (p *pipeline) NoneOf(ctx context.Context, parentResultC chan<- result, tasks []func(execCtx context.Context, resultC chan<- result)) {
+	p.execTasks(ctx, parentResultC, tasks, func(res result, isLastExpected bool) (*result, bool) {
+		// Short-circuit - return not matched
+		if res.Matched {
+			return &result{
+				Matched:      false,
+				DecisionPath: res.DecisionPath,
+				Err:          nil,
+			}, true
+		}
+		// Last result AND it's not a match due to prev condition -> return matched
+		if isLastExpected {
+			return &result{
+				Matched:      true,
+				DecisionPath: res.DecisionPath,
+				Err:          nil,
+			}, true
+		}
+		// Keep looking
+		return nil, false
 	})
-	if e != nil {
-		return false, decisionPath, nil, e
-	}
-	return match, decisionPath, newWookie, nil
 }
 
-func (svc CheckService) appendTenantContext(warrantCheck *CheckSpec, tenantId string) {
-	if warrantCheck.CheckWarrantSpec.Context == nil {
-		warrantCheck.CheckWarrantSpec.Context = warrant.PolicyContext{
-			"tenant": tenantId,
+func (p *pipeline) execTasks(ctx context.Context, parentResultC chan<- result, tasks []func(execCtx context.Context, resultC chan<- result), checkResultFunc func(r result, isLastExpected bool) (*result, bool)) {
```

Comment:
You should create type aliases for some of these common anonymous function signatures. It'll make it easier to follow and reduce some code:
```go
type taskFunc = func(...) (*result, bool)
```